modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -535,4 +604,179 @@ private static void validateParsedStatement(
+                List<ParsedResult> parsedResults,
+                Object[] params
+        ) {
+            this.schemaName = schemaName;
+            this.transactions = transactions;
+            this.explicitTransaction = explicitTransaction;
+
+            List<ScriptStatementParameters> statementsList = splitDynamicParameters(parsedResults, params);
+
+            // All statements that need to be processed.
+            statements = new ArrayBlockingQueue<>(statementsList.size(), false, statementsList);
+
+            // Statement futures that return a cursor.
+            cursorFutures = statementsList.stream()
+                    .filter(data -> data.parsedResult.queryType() != SqlQueryType.TX_CONTROL)
+                    .map(data -> data.cursorFuture)
+                    .collect(toUnmodifiableList());
+        }
+
+        AsyncSqlCursorIterator<List<Object>> asyncCursorIterator() {
+            Iterator<CompletableFuture<AsyncSqlCursor<List<Object>>>> delegate = cursorFutures.iterator();
+
+            return new AsyncSqlCursorIterator<>() {
+                @Override
+                public boolean hasNext() {
+                    return delegate.hasNext();
+                }
+
+                @Override
+                public CompletableFuture<AsyncSqlCursor<List<Object>>> next() {
+                    return delegate.next();
+                }
+            };
+        }
+
+        void processNext() {
+            ScriptStatementParameters parameters = statements.poll();
+
+            if (parameters == null) {
+                return;
+            }
+
+            ParsedResult parsedResult = parameters.parsedResult;
+            Object[] dynamicParams = parameters.dynamicParams;
+            CompletableFuture<AsyncSqlCursor<List<Object>>> cursorFuture = parameters.cursorFuture;
+
+            try {
+                // TODO https://issues.apache.org/jira/browse/IGNITE-20463
+                if (parsedResult.queryType() == SqlQueryType.TX_CONTROL) {
+                    taskExecutor.execute(this::processNext);
+                    return;
+                }
+
+                if (cursorFuture.isDone()) {
+                    return;
+                }
+
+                QueryTransactionWrapper txWrapper = wrapTxOrStartImplicit(parsedResult.queryType(), transactions, explicitTransaction);
+
+                executeParsedStatementAndWaitPrefetch(txWrapper, parsedResult, dynamicParams)
+                        .whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                txWrapper.rollback();
+                                cursorFuture.completeExceptionally(ex);
+                                cancelAll(ex);
+
+                                return;
+                            }
+
+                            if (parsedResult.queryType() == SqlQueryType.DML || parsedResult.queryType() == SqlQueryType.DDL) {
+                                txWrapper.commitImplicit();
```

Comment:
why do we commit transaction only in case of DML | DDL statements?

Reply:
:thinking:  Fixed: Implicit transaction is now committed after prefetching for all query types.